src/posix/platform/spi_interface.cpp

```diff
@@ -0,0 +1,833 @@
+    {
+        aTimeout.tv_sec  = static_cast<time_t>(timeout / kMsecPerSec);
+        aTimeout.tv_usec = static_cast<suseconds_t>((timeout % kMsecPerSec) * kUsecPerMsec);
+    }
+
+    ret = select(mIntGpioValueFd + 1, &readFdSet, NULL, NULL, &aTimeout);
+    if ((ret < 0) && (errno != EINTR))
+    {
+        DieNow(OT_EXIT_ERROR_ERRNO);
+    }
+
+    if (FD_ISSET(mIntGpioValueFd, &readFdSet))
+    {
+        struct gpioevent_data event;
+
+        otLogDebgPlat("Interrupt.");
+
+        // Read event data to clear interrupt.
+        VerifyOrDie(read(mIntGpioValueFd, &event, sizeof(event)) != -1, OT_EXIT_FAILURE);
+    }
+
+    // If we can receive a packet.
+    if (CheckInterrupt())
+    {
+        PushPullSpi();
+    }
+
+    return error;
+}
+
+otError SpiInterface::SendFrame(const uint8_t *aFrame, uint16_t aLength)
+{
+    otError error = OT_ERROR_NONE;
+
+    VerifyOrExit(aLength < (kMaxFrameSize - kSpiFrameHeaderSize), error = OT_ERROR_NO_BUFS);
+    VerifyOrExit(!mSpiTxIsReady, error = OT_ERROR_BUSY);
+
+    memcpy(&mSpiTxFrameBuffer[kSpiFrameHeaderSize], aFrame, aLength);
+
+    mSpiTxIsReady     = true;
+    mSpiTxPayloadSize = aLength;
+
+    PushPullSpi();
+
+exit:
+    return error;
+}
+
+void SpiInterface::HandleReceivedFrame(Ncp::SpiFrame &aSpiFrame)
+{
+    const uint8_t *spinelFrame = aSpiFrame.GetData();
+
+    for (uint16_t i = 0; i < aSpiFrame.GetHeaderDataLen(); i++)
+    {
+        if (mRxFrameBuffer.WriteByte(spinelFrame[i]) != OT_ERROR_NONE)
+        {
+            mRxFrameBuffer.DiscardFrame();
+            otLogNotePlat("No enough memory buffers, drop packet");
+            ExitNow();
+        }
+    }
+
+    mCallbacks.HandleReceivedFrame();
+
+exit:
+    return;
+}
+
+void SpiInterface::LogError(const char *aString)
+{
+    fprintf(stderr, "%s: %s", aString, strerror(errno));
+    otLogCritPlat("%s: %s", aString, strerror(errno));
+}
+
+void SpiInterface::LogStats(void)
+{
+    otLogInfoPlat("INFO: mSlaveResetCount=%llu", (unsigned long long)mSlaveResetCount);
+    otLogInfoPlat("INFO: mSpiFrameCount=%llu", (unsigned long long)mSpiFrameCount);
+    otLogInfoPlat("INFO: mSpiValidFrameCount=%llu", (unsigned long long)mSpiValidFrameCount);
+    otLogInfoPlat("INFO: mSpiDuplexFrameCount=%llu", (unsigned long long)mSpiDuplexFrameCount);
+    otLogInfoPlat("INFO: mSpiUnresponsiveFrameCount=%llu", (unsigned long long)mSpiUnresponsiveFrameCount);
+    otLogInfoPlat("INFO: mSpiGarbageFrameCount=%llu", (unsigned long long)mSpiGarbageFrameCount);
+    otLogInfoPlat("INFO: mSpiRxFrameCount=%llu", (unsigned long long)mSpiRxFrameCount);
+    otLogInfoPlat("INFO: mSpiRxFrameByteCount=%llu", (unsigned long long)mSpiRxFrameByteCount);
+    otLogInfoPlat("INFO: mSpiTxFrameCount=%llu", (unsigned long long)mSpiTxFrameCount);
+    otLogInfoPlat("INFO: mSpiTxFrameByteCount=%llu", (unsigned long long)mSpiTxFrameByteCount);
+}
+
+#if OPENTHREAD_CONFIG_LOG_LEVEL >= OT_LOG_LEVEL_WARN
```

Comment:
posix-host uses dynamic logging. Use runtime check instead?

Reply:
This function has already been removed. We don't need this logging anymore.